cleverhans/future/tf2/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,334 @@
+    :param y: (optional) Targeted attack? 
+    :param batch_size: Number of attacks to run simultaneously.
+    :param clip_min: (optional) float. Minimum float values for adversarial example components.
+    :param clip_max: (optional) float. Maximum float value for adversarial example components.
+    :param binary_search_steps: The number of times we perform binary
+                            search to find the optimal tradeoff-
+                            constant between norm of the purturbation
+                            and confidence of the classification.
+    :param max_iterations: The maximum number of iterations. Setting this
+                           to a larger value will produce lower distortion
+                           results. Using only a few iterations requires
+                           a larger learning rate, and will produce larger
+                           distortion results.
+    :param abort_early: If true, allows early aborts if gradient descent
+                    is unable to make progress (i.e., gets stuck in
+                    a local minimum).
+    :param confidence: Confidence of adversarial examples: higher produces
+                       examples with larger l2 distortion, but more
+                       strongly classified as adversarial.
+    :param initial_const: The initial tradeoff-constant to use to tune the
+                      relative importance of size of the pururbation
+                      and confidence of classification.
+                      If binary_search_steps is large, the initial
+                      constant is not important. A smaller value of
+                      this constant gives lower distortion results.
+    :param learning_rate: The learning rate for the attack algorithm.
+                      Smaller values produce better results but are
+                      slower to converge.
+    """
+    self.model_fn = model_fn
+
+    self.batch_size = batch_size
+
+    self.y = y
+    self.targeted = y is not None
+
+    self.clip_min = clip_min
+    self.clip_max = clip_max
+
+    self.binary_search_steps = binary_search_steps
+    self.max_iterations = max_iterations
+    self.abort_early = abort_early
+    self.learning_rate = learning_rate
+
+    self.confidence = confidence
+    self.initial_const = initial_const
+
+    # the optimizer
+    self.optimizer = tf.keras.optimizers.Adam(self.learning_rate)
+
+    super(CarliniWagnerL2, self).__init__()
+
+  def attack(self, x):
+    """
+    Returns adversarial examples for the tensor.
+    :param x: input tensor.
+    :return: a numpy tensor with the adversarial example.
+    """
+    adv_ex = np.zeros_like(x)
+    for i in range(0, len(x), self.batch_size):
+      adv_ex[i:i +
+             self.batch_size] = self._attack(x[i:i+self.batch_size]).numpy()
+
+    return adv_ex
+
+  def _attack(self, x):
+    if self.clip_min is not None:
+      assert np.all(tf.math.greater_equal(x, self.clip_min))
```

Comment:
I would remove those assertion in final code, I feel like they are redundant with checks already embedded in the library. Or maybe turn them into a custom error?

Reply:
Addressed in 0a847c9